pkg/actuators/machineset/controller.go

```diff
@@ -0,0 +1,127 @@
+	// This is needed by the autoscaler to foresee upcoming capacity when scaling from zero.
+	// https://github.com/openshift/enhancements/pull/186
+	cpuKey    = "machine.openshift.io/vCPU"
+	memoryKey = "machine.openshift.io/memoryMb"
+	gpuKey    = "machine.openshift.io/GPU"
+)
+
+// Reconciler reconciles machineSets.
+type Reconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&machinev1.MachineSet{}).
+		WithOptions(options).
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("machineset-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+// Reconcile implements controller runtime Reconciler interface.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("machineset", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	ctx := context.Background()
+	machineSet := &machinev1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineSets, this can happen when foregroundDeletion
+	// is enabled
+	if !machineSet.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	originalMachineSetToPatch := client.MergeFrom(machineSet.DeepCopy())
+
+	result, err := reconcile(machineSet)
+	if err != nil {
+		logger.Error(err, "Failed to reconcile MachineSet")
+		r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineSet, originalMachineSetToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineSet: %v", err)
+	}
+
+	return result, err
+}
+
+func reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, error) {
+	providerConfig, err := getProviderConfig(*machineSet)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get providerConfig: %v", err)
+	}
+	instanceType, ok := InstanceTypes[providerConfig.InstanceType]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("unknown instance type: %s", providerConfig.InstanceType)
+	}
+
+	if machineSet.Annotations == nil {
+		machineSet.Annotations = make(map[string]string)
+	}
+
+	// TODO: get annotations keys from machine API
+	machineSet.Annotations[cpuKey] = strconv.FormatInt(instanceType.VCPU, 10)
```

Comment:
Should we overwrite the annotations if they already exist (eg, user overrides them for one reason or another)?  It's conceivable that for one reason or another they might need to specify different values.  This is probably not a blocker for the first iteration, but possibly a TODO item for follow up.

Reply:
This is a good point. I wouldn't add that complexity until there're cases which justify it